Support Python 3.12

jgraham opened the pull request:
This requires a patch to pywebsocket3, since it doesn't correctly use the `SSLContext` and the old approach has now been dropped. We can upstream this patch after landing here, but this is quite urgent now since several major platforms are shipping 3.12 by default.

It also updates all the CI configuration to use Python 3.12 in addition to, or instead of, Python 3.11, as appropriate.

jgraham committed:
Fix pywebsocket3 to support Python 3.12

One can no longer just call ssl.wrap_socket, but has to set up an SSLContext object
with the required options.

jgraham committed:
Update Azure jobs to Python 3.12

jgraham committed:
Update Taskcluster jobs to Python 3.12

jgraham committed:
Update GitHub actions workflows to use Python 3.12

jgraham committed:
Update all tox configuration to use Python 3.12

foolip reviewed (approved):
Seems like more changes are needed to pass CI, but the changes so far make sense. If only trivial changes are needed then I don't need to review again.

jgraham committed:
Stop using datetime.utcnow

Instead use datetime.now(timezone.utc) to create a timeszone-aware
datetime object.

jgraham committed:
Update pillow version

The older release doesn't support 3.12

jgraham committed:
Update to latest zstandard

jgraham committed:
Handle tarfile.extractall warning

Fix utils.untar to set the filter argument for newer python versions, and ignore
the warning coming from mozfile

jgraham committed:
Don't install distutils package for Python 3.12

This makes the CI environment closer to user environments.

jgraham committed:
Update to latest mozdevice
